### PR TITLE
Expand bi-directional associations guide [ci skip]

### PR DESCRIPTION
**Pull request by composerinteralia**
I remember being confused about what exactly association inverses would
do for me. This guide helped, but it leaves out some use cases. I'm hoping
that having all these use cases documented in one place with code
examples for each will make it easier for folks to understand why
they might want to set `:inverse_of` when Rails can't infer it.

Specifically, this commit:

* Adds an example explicitly demonstrating that all the `Author` objects are the same, and that no additional queries are executed.
* Preserves the example around preventing inconsistent data (although replaces the single-letter variable names)
* Adds an example around autosaving, which I'm not sure is explicitly documented elsewhere
* Adds an example around validations, linking to the relevant sections of the validation guide where we mention needing to set the `:inverse_of` for these validations

This commit also adds links from the `:inverse_of` sections back up to
the Bi-directional associations sections, hopefully making those
`:inverse_of` sections clearer.

### CODE REVIEW

File: guides/source/association_basics.md [hunk 1]
```diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
...
-Active Record will attempt to automatically identify that these two models share a bi-directional association based on the association name. In this way, Active Record will only load one copy of the `Author` object, making your application more efficient and preventing inconsistent data:
+Active Record will attempt to automatically identify that these two models share
+a bi-directional association based on the association name. This information
+allows Active Record to:
+
+* Prevent needless queries for already-loaded data
+
+```irb
+irb> author = Author.first
+irb> authors = author.books.all? do |book|
```
Comment (kamipo) [inline guides/source/association_basics.md:772]:
```suggestion
irb> author.books.all? do |book|
```

Comment (composerinteralia) [reply guides/source/association_basics.md]: Nice catch!

File: guides/source/association_basics.md [hunk 2]
```diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
...
+
+* Prevent needless queries for already-loaded data
+
+```irb
+irb> author = Author.first
+irb> author.books.all? do |book|
+irb>   book.author.equal?(author) # No additional queries executed here
+irb> end
+=> true
+```
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:776]: If we want to maintain the bulleted list margin, we should indent these code snippets:

```suggestion
    ```irb
    irb> author = Author.first
    irb> author.books.all? do |book|
    irb>   book.author.equal?(author) # No additional queries executed here
    irb> end
    => true
    ```
```

File: guides/source/association_basics.md [hunk 3]
```diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
...
+irb> author = Author.new
+irb> book = author.books.new
+irb> book.save!
+irb> book.persisted?
+=> true
+irb> author.persisted?
+=> true
+```
+
+* Validate the [presence](active_record_validations.html#presence) and [absence](active_record_validations.html#absence) of associations in more cases
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:803]: I don't have a strong opinion about wrapping markdown, but, for consistency with the rest of this section:

```suggestion
* Validate the [presence](active_record_validations.html#presence) and 
  [absence](active_record_validations.html#absence) of associations in more cases
```

File: guides/source/association_basics.md [hunk 4]
```diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
...
+
+```irb
+irb> author = Author.new
+irb> book = author.books.new
+irb> book.author.present?
+=> true
+irb> book.valid?
 => true
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:811]: Perhaps a counter example here, to further illustrate the point?

```suggestion
irb> book = Book.new
irb> book.author.present?
=> false
irb> book.valid?
=> false
irb> author = Author.new
irb> book = author.books.new
irb> book.author.present?
=> true
irb> book.valid?
=> true
```

Or maybe with validation error messages instead?

```suggestion
irb> book = Book.new
irb> book.valid?
=> false
irb> book.errors.full_messages
=> ["Author must exist"]
irb> author = Author.new
irb> book = author.books.new
irb> book.valid?
=> true
```

File: guides/source/association_basics.md [hunk 5]
```diff
@@ -761,15 +761,53 @@ class Book < ApplicationRecord
...
-irb> a = Author.first
-irb> b = a.books.first
-irb> a.first_name == b.author.first_name
+irb> author = Author.first
+irb> book = author.books.first
+irb> author.name == book.author.name
 => true
-irb> a.first_name = 'David'
-irb> a.first_name == b.author.first_name
+irb> author.name = 'Changed Name'
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:786]:
```suggestion
irb> author.name = "Changed Name"
```

File: guides/source/association_basics.md [hunk 6]
```diff
@@ -793,15 +831,51 @@ class Book < ApplicationRecord
...
+```
+
+* Reference multiple copies of a model with inconsistent data
+
+```irb
+irb> author = Author.first
+irb> book = author.books.first
+irb> author.name == book.author.name
+=> true
+irb> author.name = 'Changed Name'
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:854]:
```suggestion
irb> author.name = "Changed Name"
```

File: guides/source/association_basics.md [hunk 7]
```diff
@@ -793,15 +831,51 @@ class Book < ApplicationRecord
...
+
+* Fail to validate presence or absence
+
+```irb
+irb> author = Author.new
+irb> book = author.books.new
+irb> book.author.present?
+=> false
+irb> book.valid?
 => false
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:879]: If you do update the snippet from line 806 to show validation error messages, this snippet should be updated to match.

**Review by jonathanhefner — COMMENTED** [review]
Nice! :+1:

**Comment by composerinteralia** [comment]
Thank you! Updated to include these suggestions.

File: guides/source/association_basics.md [hunk 8]
```diff
@@ -761,17 +761,60 @@ class Book < ApplicationRecord
...
+
+* Prevent needless queries for already-loaded data
+
+  ```irb
+  irb> author = Author.first
+  irb> author.books.all? do |book|
+  irb>   book.author.equal?(author) # No additional queries executed here
+  irb> end
+  => true
+  ```
```
Comment (jonathanhefner) [inline guides/source/association_basics.md:776]: Unfortunately, markdown doesn't recognize the code block as being part of the list item if it only has a 2-space indent.  It has to be at least 4 spaces:

```suggestion
    ```irb
    irb> author = Author.first
    irb> author.books.all? do |book|
    irb>   book.author.equal?(author) # No additional queries executed here
    irb> end
    => true
    ```
  ```
```

Comment (composerinteralia) [reply guides/source/association_basics.md]: 🤦🏻 Thanks for spotting that!

**Comment by jonathanhefner** [comment]
Thank you, @composerinteralia! :wave:

**Pull request merged**